ignite/metrics/nlp/bleu.py

```diff
@@ -146,59 +166,128 @@ def _corpus_bleu(
+        return (p_numerators, p_denominators, hyp_lengths, ref_lengths)
+
+    @staticmethod
+    def _brevity_penalty_smoothing(
+        p_numerators: Counter,
+        p_denominators: Counter,
+        smoother: _Smoother,
+        hyp_length_sum: int,
+        ref_length_sum: int,
+        weights: Sequence[float],
+    ) -> float:
+
         # Returns 0 if there's no matching n-grams
         # We only need to check for p_numerators[1] == 0, since if there's
         # no unigrams, there won't be any higher order ngrams.
         if p_numerators[1] == 0:
             return 0
 
         # If no smoother, returns 0 if there's at least one a not matching n-grams
-        if self.smoother.smooth == "no_smooth" and min(p_numerators.values()) == 0:
+        if smoother.smooth == "no_smooth" and min(p_numerators.values()) == 0:
             return 0
 
-        # Calculate the hypothesis lengths
-        hyp_lengths = [len(hyp) for hyp in candidates]
-
-        # Calculate the closest reference lengths.
-        ref_lengths = [_closest_ref_length(refs, hyp_len) for refs, hyp_len in zip(references, hyp_lengths)]
-
-        # Sum of hypothesis and references lengths
-        hyp_len = sum(hyp_lengths)
-        ref_len = sum(ref_lengths)
-
         # Calculate corpus-level brevity penalty.
-        if hyp_len < ref_len:
-            bp = math.exp(1 - ref_len / hyp_len) if hyp_len > 0 else 0.0
+        if hyp_length_sum < ref_length_sum:
+            bp = math.exp(1 - ref_length_sum / hyp_length_sum) if hyp_length_sum > 0 else 0.0
         else:
             bp = 1.0
-
         # Smoothing
-        p_n = self.smoother(p_numerators, p_denominators)
+        p_n = smoother(p_numerators, p_denominators)
 
         # Compute the geometric mean
-        s = [w_i * math.log(p_i) for w_i, p_i in zip(self.weights, p_n)]
+        s = [w_i * math.log(p_i) for w_i, p_i in zip(weights, p_n)]
         gm = bp * math.exp(math.fsum(s))
         return gm
 
+    def _sentence_bleu(
+        self, references: Sequence[Sequence[Sequence[Any]]], candidates: Sequence[Sequence[Any]],
+    ) -> float:
+        p_numerators: Counter = Counter()
+        p_denominators: Counter = Counter()
+        p_numerators, p_denominators, hyp_lengths, ref_lengths = self._n_gram_counter(
+            references=references,
+            candidates=candidates,
+            p_numerators=p_numerators,
+            p_denominators=p_denominators,
+            ngrams_order=self.ngrams_order,
+        )
+        bleu_score = self._brevity_penalty_smoothing(
+            p_numerators=p_numerators,
+            p_denominators=p_denominators,
+            smoother=self.smoother,
+            hyp_length_sum=sum(hyp_lengths),
+            ref_length_sum=sum(ref_lengths),
+            weights=self.weights,
+        )
+        return bleu_score
+
+    def _corpus_bleu(self, references: Sequence[Sequence[Sequence[Any]]], candidates: Sequence[Sequence[Any]],) -> None:
```

Comment:
I don’t understand this implementation.

IMO what is done in `_sentence_bleu` should be moved here. And after thinking about it, `_sentence_bleu` should call `_corpus_bleu` on the unique input sentence. See the implementation in nltk. 

https://github.com/nltk/nltk/blob/f989fe65d421e7ea4d1037a00f07eaeee3ad6a29/nltk/translate/bleu_score.py#L97

My point is, should we provide micro and macro corpus bleu ?

Reply:
Fixed this and added tests also, please take a look